### PR TITLE
build: Update leveldb to 1.22+

### PR DESCRIPTION
**Pull request by laanwj**
This updates leveldb to currently newest upstream commit https://github.com/bitcoin-core/leveldb/commit/0c40829872a9f00f38e11dc370ff8adb3e19f25b:

- CRC32C hardware acceleration is now an external library [crc32c](https://github.com/google/crc32c). This adds acceleration on ARM, and should be faster on x86 because of using prefetch. It also makes it easy to support similar instruction sets on other platforms in the future.
- Thread handling uses C++11, instead of platform specific code.
- Native windows environment was added. No need to maintain our own hacky one, anymore.
- Upstream now builds using CMake. This doesn't mean we need to use that (phew), but internal configuration changed to a a series of checks, instead of OS profiles. This means the blanket error "Cannot build leveldb for $host. Please file a bug report' is removed.

All changes: https://github.com/google/leveldb/compare/a53934a3ae1244679f812d998a4f16f2c7f309a6...0c40829872a9f00f38e11dc370ff8adb3e19f25b

Pretty much all our changes have been subsumed by upstream, so we figured it was cleaner to start over with a new branch from upstream with the still-relevant patches applied: https://github.com/bitcoin-core/leveldb/tree/bitcoin-fork-new

There's quite some testing to be done (see below). See https://github.com/bitcoin-core/leveldb/issues/25 and https://github.com/bitcoin-core/leveldb/pull/26 for more history and context.

TODO:
- [x] Subtree `crc32c`
- [x] Make linters happy about crc32 subtree
- [x] Integrate `crc32c` library into build system
- [x] MSVC build system

### CODE REVIEW

**Comment by laanwj** [comment]
OK, travis build should pass now (even windows, though not the MSVC build system). Testing notes:

- [ ] check that `HAVE_FULLFSYNC=1` is passed on MacOS
- [ ] test on BSD variants
- [ ] test on Windows (full sync, as well as catching up an existing node)

**Comment by elichai** [comment]
Concept ACK for the new branch. it's really nice that we can again easily see the diff from upstream.
https://github.com/google/leveldb/compare/master...bitcoin-core:bitcoin-fork-new

Will try to read later the (reported) changes upstream leveldb did in the past ~160 commits.

**Comment by sipsorcery** [comment]
Regarding the msvc build I didn't have any problems getting it to build after applying what I think are the relevant changes to the `libleveldb` project.

I do get two new warnings:

````
Severity	Code	Description	Project	File	Line	Suppression State	Suppression State
Warning	C26110	Caller failing to hold lock 'this->mu_' before calling function 'std::_Mutex_base::unlock'.	libleveldb	C:\DEV\GITHUB\SIPSORCERY_BITCOIN\SRC\LEVELDB\PORT\PORT_STDCXX.H	58		
Severity	Code	Description	Project	File	Line	Suppression State	Suppression State
Warning	C26110	Caller failing to hold lock 'this->mu_->mu_' before calling function 'std::unique_lock<std::mutex>::unique_lock<std::mutex>'.	libleveldb	C:\DEV\GITHUB\SIPSORCERY_BITCOIN\SRC\LEVELDB\PORT\PORT_STDCXX.H	76		
````

Are any of the preprocessor defines below relevant for Windows?

````
LEVELDB_CPPFLAGS_INT += -DHAVE_FDATASYNC=@HAVE_FDATASYNC@
LEVELDB_CPPFLAGS_INT += -DHAVE_FULLFSYNC=@HAVE_FULLFSYNC@
LEVELDB_CPPFLAGS_INT += -DHAVE_O_CLOEXEC=@HAVE_O_CLOEXEC@
````

In case it's of any use the diff for `libleveldb.vcxproj` is:

````
diff --git a/build_msvc/libleveldb/libleveldb.vcxproj b/build_msvc/libleveldb/libleveldb.vcxproj
index f855923c6..ba6a86fdc 100644
--- a/build_msvc/libleveldb/libleveldb.vcxproj
+++ b/build_msvc/libleveldb/libleveldb.vcxproj
@@ -24,8 +24,6 @@
     <ClCompile Include="..\..\src\leveldb\db\version_set.cc" />
     <ClCompile Include="..\..\src\leveldb\db\write_batch.cc" />
     <ClCompile Include="..\..\src\leveldb\helpers\memenv\memenv.cc" />
-    <ClCompile Include="..\..\src\leveldb\port\port_posix_sse.cc" />
-    <ClCompile Include="..\..\src\leveldb\port\port_win.cc" />
     <ClCompile Include="..\..\src\leveldb\table\block.cc" />
     <ClCompile Include="..\..\src\leveldb\table\block_builder.cc" />
     <ClCompile Include="..\..\src\leveldb\table\filter_block.cc" />
@@ -42,7 +40,7 @@
     <ClCompile Include="..\..\src\leveldb\util\comparator.cc" />
     <ClCompile Include="..\..\src\leveldb\util\crc32c.cc" />
     <ClCompile Include="..\..\src\leveldb\util\env.cc" />
-    <ClCompile Include="..\..\src\leveldb\util\env_win.cc" />
+    <ClCompile Include="..\..\src\leveldb\util\env_windows.cc" />
     <ClCompile Include="..\..\src\leveldb\util\filter_policy.cc" />
     <ClCompile Include="..\..\src\leveldb\util\hash.cc" />
     <ClCompile Include="..\..\src\leveldb\util\histogram.cc" />
@@ -52,7 +50,7 @@
   </ItemGroup>
   <ItemDefinitionGroup>
     <ClCompile>
-        <PreprocessorDefinitions>_CRT_NONSTDC_NO_DEPRECATE;LEVELDB_PLATFORM_WINDOWS;LEVELDB_ATOMIC_PRESENT;%(PreprocessorDefinitions)</PreprocessorDefinitions>
+        <PreprocessorDefinitions>HAVE_CRC32C=0;HAVE_SNAPPY=0;__STDC_LIMIT_MACROS;LEVELDB_IS_BIG_ENDIAN=0;_UNICODE;UNICODE;_CRT_NONSTDC_NO_DEPRECATE;LEVELDB_PLATFORM_WINDOWS;LEVELDB_ATOMIC_PRESENT;%(PreprocessorDefinitions)</PreprocessorDefinitions>
         <DisableSpecificWarnings>4244;4267;4312;</DisableSpecificWarnings>
         <AdditionalIncludeDirectories>..\..\src\leveldb;..\..\src\leveldb\include;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
        </ClCompile>
````

**Comment by laanwj** [comment]
> In case it's of any use the diff for libleveldb.vcxproj is:

Thanks, I'll include that!

> Are any of the preprocessor defines below relevant for Windows?

I think they're only used in env_posix.cc? if so, no

**Comment by laanwj** [comment]
AppVeyor fails on some warning that is promoted to error, for an unreachable destructor.
```
C:\projects\bitcoin\src\leveldb\util\env_windows.cc(383): error C2220: the following warning is treated as an error [C:\projects\bitcoin\build_msvc\bitcoind\bitcoind.vcxproj]
C:\projects\bitcoin\src\leveldb\util\env_windows.cc(383): warning C4722: 'leveldb::`anonymous namespace'::WindowsEnv::~WindowsEnv': destructor never returns, potential memory leak [C:\projects\bitcoin\build_msvc\bitcoind\bitcoind.vcxproj]
```
Did you need some flag changes too?

**Comment by fanquake** [comment]
Concept ACK - have not reviewed the changes yet. Only done some minimal does it compile + start running tests. Started testing going between leveldb versions on macOS.

### macOS (fed6a9bd661321694b375dc393e8c0307042474f + depends):
```bash
checking for fdatasync... no
checking for F_FULLFSYNC... yes
checking for O_CLOEXEC... yes
checking for __builtin_prefetch... yes
checking for _mm_prefetch... yes
checking for strong getauxval support in the system headers... no
checking for weak getauxval support in the compiler... yes
....
LevelDB using max_open_files=1000 (default=1000)
```

### openBSD 6.6 (fed6a9bd661321694b375dc393e8c0307042474f)
```bash
checking for fdatasync... yes
checking for F_FULLFSYNC... no
checking for O_CLOEXEC... yes
checking for __builtin_prefetch... yes
checking for _mm_prefetch... yes
checking for strong getauxval support in the system headers... no
checking for weak getauxval support in the compiler... yes
....
LevelDB using max_open_files=1000 (default=1000)
```

### Windows 10 (WSL) (fed6a9bd661321694b375dc393e8c0307042474f + depends)
```bash
checking for fdatasync... no
checking for F_FULLFSYNC... no
checking for O_CLOEXEC... no
checking for __builtin_prefetch... yes
checking for _mm_prefetch... yes
checking for strong getauxval support in the system headers... no
checking for weak getauxval support in the compiler... yes
...
LevelDB using max_open_files=1000 (default=1000)
```

**Comment by laanwj** [comment]
@fanquake thanks for testing, good to hear that this is not causing build failure anywhere

**Comment by laanwj** [comment]
Squashed fixup commits and removed WIP tag. Should be ready for review/testing.

**Comment by sipa** [comment]
ACK.

Concept: very happy to get rid of all those local modifications.

Code review: I reviewed the diff between upstream LevelDB and our new branch (but not the changes in upstream LevelDB), and the new commits here.

Tests:
* I verified that the crc32 asm code is being executed on x86_64 by adding asserts before the asm commands in the crc32 lib.
* I took a datadir that was a few weeks out of date on an x86_64 machine, started it with this branch, and ran with -checkblocks=500. No issues.
* I let this datadir sync with the network, interrupting it from time to time to force a flush.
* I ran gettxoutsetinfo on this branch on the resulting datadir, and compared it with a node that has been running master as of a few weeks.
* I restarted with master, let it sync another 3 blocks, and ran gettxoutsetinfo again, and compared it again with another node.
* I then let that datadir sync

**Comment by sipsorcery** [comment]
tACK* e2ef5fbc0ac077f942aa79dfb4274e0016701c24.

*test on Windows (full sync, as well as catching up an existing node)

**Comment by sipa** [comment]
Is commit https://github.com/bitcoin/bitcoin/commit/abd50f18e80f64f35cb3b6747b8d1f7e6617d977 still needed?

**Comment by laanwj** [comment]
> Is commit abd50f1 still needed?

The subtree check in travis didn't pass with it. I could try removing it, but I think that's still the case. 

I don't understand how it's supposed to pass in the first place for the other dependencies without ever fetching them as described in the [doc](https://github.com/bitcoin/bitcoin/blob/master/test/lint/README.md#git-subtree-checksh)? Or does travis do this hidden?

**Comment by MarcoFalke** [comment]
We only check that it is a subtree, not that the subtree corresponds to some subtree hosted somewhere else.

**Comment by laanwj** [comment]
> We only check that it is a subtree, not that the subtree corresponds to some subtree hosted somewhere else.

Are you sure? Now I'm doubly confused why that commit makes a difference at all. Did I subtree correctly?

**Comment by MarcoFalke** [comment]
Maybe the script doesn't work for initial subtree check ins?

**Comment by laanwj** [comment]
I don't understand why that would be the case. What is the difference between this tree, and the tree when it is merged into master, that would make it pass a subtree check? In both cases the merge of `crc32c` is simply an ancestor, it's one merge deeper then.

**Comment by MarcoFalke** [comment]
I removed the commit and travis passed for me :woman_shrugging:

**Comment by laanwj** [comment]
oh great, thanks, have i been chasing ghosts again… removed it here too

**Comment by fanquake** [comment]
@theuni Cory it would be good to get your eyes over the build system changes here.

**Comment by MarcoFalke** [comment]
@jamesob Is this something you could benchmark?

**Comment by jamesob** [comment]
Sure thing, I'll start a benchmark run this afternoon.

**Comment by laanwj** [comment]
Travis still passes after removing the subtree workaround (yay). AppVeyor is still failing on [this issue](https://github.com/bitcoin/bitcoin/pull/17398#issuecomment-551748307):

> AppVeyor fails on some warning that is promoted to error, for an unreachable destructor.

Anyone have ideas how to disable this warning, or at least make it no longer an error?

**Comment by sipsorcery** [comment]
@laanwj msvc is grumpy about the `std::abort` in the destructor. It does seem like the nuclear option but I guess the leveldb devs decided the condition is catastrophic enough to warrant it.

To get rid of the warning and get the appveyor build to pass the line in `build_msvc\common.init.vcxproj` file can be updated to:

````
<DisableSpecificWarnings>2220;4018;4221;4244;4267;4334;4715;4805;</DisableSpecificWarnings>
````

**Comment by laanwj** [comment]
> @laanwj msvc is grumpy about the std::abort in the destructor. It does seem like the nuclear option but I guess the leveldb devs decided the condition is catastrophic enough to warrant it.

I would also prefer if they do something different, but I'll leave that for upstream.

> To get rid of the warning and get the appveyor build to pass the line in build_msvc\common.init.vcxproj file can be updated to:

Thank you, will add this.

**Comment by sipsorcery** [comment]
Hmm leave that appveyor error with me. I'll track down why the error code reported isn't accepted in the ignore list.

**Comment by sipa** [comment]
Perhaps you want disable C4722 (destructor never terminates), not C2220 (treat warnings as errors)?

**Comment by sipsorcery** [comment]
@sipa yes, I should have read the trace more closely. Double checking build with corrected fix at the moment.

**Comment by sipa** [comment]
@laanwj Just discovered that the new native windows env here also has support for mmap'ed files, which may potentially be a significant performance improvement for windows systems. It's also limited to 1000 though. I don't think we want to change that without experimentation, but thought to bring it up so that we're aware of it.

**Comment by MarcoFalke** [comment]
Would it make sense to create the crc32c subtree in `src/subtree/` instead of in `src/`? That way it makes it doubly clear that it is a subtree and in light of #17413 it wouldn't have to be moved again when someone works on that issue.

**Comment by sipsorcery** [comment]
Diff to fix the appveyor build. Removes the previously suggested warning fix and puts the correct one into the `leveldb` project.

````
diff --git a/build_msvc/common.init.vcxproj b/build_msvc/common.init.vcxproj
index ddd34603c..a04a38ff7 100644
--- a/build_msvc/common.init.vcxproj
+++ b/build_msvc/common.init.vcxproj
@@ -107,7 +107,7 @@
       <WarningLevel>Level3</WarningLevel>
       <PrecompiledHeader>NotUsing</PrecompiledHeader>
       <AdditionalOptions>/utf-8 %(AdditionalOptions)</AdditionalOptions>
-      <DisableSpecificWarnings>2220;4018;4221;4244;4267;4334;4715;4805;</DisableSpecificWarnings>
+      <DisableSpecificWarnings>4018;4221;4244;4267;4334;4715;4805;</DisableSpecificWarnings>
       <TreatWarningAsError>true</TreatWarningAsError>
       <PreprocessorDefinitions>ZMQ_STATIC;NOMINMAX;WIN32;HAVE_CONFIG_H;_CRT_SECURE_NO_WARNINGS;_SCL_SECURE_NO_WARNINGS;_CONSOLE;_WIN32_WINNT=0x0601;%(PreprocessorDefinitions)</PreprocessorDefinitions>
       <AdditionalIncludeDirectories>..\..\src;..\..\src\univalue\include;..\..\src\secp256k1\include;..\..\src\leveldb\include;..\..\src\leveldb\helpers\memenv;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
diff --git a/build_msvc/libleveldb/libleveldb.vcxproj b/build_msvc/libleveldb/libleveldb.vcxproj
index cabc17755..fc2bca12b 100644
--- a/build_msvc/libleveldb/libleveldb.vcxproj
+++ b/build_msvc/libleveldb/libleveldb.vcxproj
@@ -50,10 +50,10 @@
   </ItemGroup>
   <ItemDefinitionGroup>
     <ClCompile>
-        <PreprocessorDefinitions>HAVE_CRC32C=0;HAVE_SNAPPY=0;__STDC_LIMIT_MACROS;LEVELDB_IS_BIG_ENDIAN=0;_UNICODE;UNICODE;_CRT_NONSTDC_NO_DEPRECATE;LEVELDB_PLATFORM_WINDOWS;LEVELDB_ATOMIC_PRESENT;%(PreprocessorDefinitions)</PreprocessorDefinitions>
-        <DisableSpecificWarnings>4244;4267;4312;</DisableSpecificWarnings>
-        <AdditionalIncludeDirectories>..\..\src\leveldb;..\..\src\leveldb\include;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
-       </ClCompile>
+     <PreprocessorDefinitions>HAVE_CRC32C=0;HAVE_SNAPPY=0;__STDC_LIMIT_MACROS;LEVELDB_IS_BIG_ENDIAN=0;_UNICODE;UNICODE;_CRT_NONSTDC_NO_DEPRECATE;LEVELDB_PLATFORM_WINDOWS;LEVELDB_ATOMIC_PRESENT;%(PreprocessorDefinitions)</PreprocessorDefinitions>
+     <DisableSpecificWarnings>4244;4267;4312;4722;</DisableSpecificWarnings>
+     <AdditionalIncludeDirectories>..\..\src\leveldb;..\..\src\leveldb\include;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
+    </ClCompile>
   </ItemDefinitionGroup>
   <Import Project="$(VCTargetsPath)\Microsoft.Cpp.props" />
   <Import Project="$(VCTargetsPath)\Microsoft.Cpp.targets" />
````

**Comment by laanwj** [comment]
> Just discovered that the new native windows env here also has support for mmap'ed files, 

Very good news!

> It's also limited to 1000 though. I don't think we want to change that without experimentation

Agree, this can be changed later, but at least for the initial merge of this, I think we'd want to stick with the default here.

> Would it make sense to create the crc32c subtree in src/subtree/ instead of in src/?

Could do that, although by now it's kind of annoying to update all the changes here, especially the linter changes. And making `src/subtree` doesn't make sense if there's no actual plans to move other subtrees there. Do you agree with moving leveldb there too?

(If we do this, I'd slightly prefer `thirdparty` to `subtrees`).

> and in light of #17413 it wouldn't have to be moved again when someone works on that issue.

I had the impression that people weren't enthusiastic about actually moving subtrees around, so I settled with "a central exception list" for all linters and scripts.

**Comment by sipa** [comment]
I like the idea of a subtrees/thirdparty/whatever subdirectory, but not as part of this PR.

**Comment by jamesob** [comment]
FWIW, 1b2407c97b4bd033efcbf6bb239cf3affbe696eb fails to compile for me:
```
./configure LDFLAGS=-L/home/james/src/bitcoin/db4/lib/ CPPFLAGS=-I/home/james/src/bitcoin/db4/include/ CXXFLAGS=-std=c++11 --enable-wallet --enable-debug --with-daemon CXX=/usr/bin/g++
[...]
  CXXLD    libunivalue.la
In file included from ./leveldb/port/port.h:14:0,
                 from ./leveldb/db/filename.h:16,
                 from leveldb/db/filename.cc:5:
./leveldb/port/port_stdcxx.h:44:36: error: ‘LEVELDB_IS_BIG_ENDIAN’ was not declared in this scope
 static const bool kLittleEndian = !LEVELDB_IS_BIG_ENDIAN;
                                    ^~~~~~~~~~~~~~~~~~~~~
./leveldb/port/port_stdcxx.h:44:36: note: suggested alternative: ‘__BIG_ENDIAN’
 static const bool kLittleEndian = !LEVELDB_IS_BIG_ENDIAN;
                                    ^~~~~~~~~~~~~~~~~~~~~
                                    __BIG_ENDIAN
Makefile:6624: recipe for target 'leveldb/db/leveldb_libleveldb_a-filename.o' failed
make[2]: *** [leveldb/db/leveldb_libleveldb_a-filename.o] Error 1
make[2]: *** Waiting for unfinished jobs....
In file included from ./leveldb/port/port.h:14:0,
                 from ./leveldb/util/coding.h:18,
                 from leveldb/db/log_reader.cc:10:
./leveldb/port/port_stdcxx.h:44:36: error: ‘LEVELDB_IS_BIG_ENDIAN’ was not declared in this scope
 static const bool kLittleEndian = !LEVELDB_IS_BIG_ENDIAN;
                                    ^~~~~~~~~~~~~~~~~~~~~
./leveldb/port/port_stdcxx.h:44:36: note: suggested alternative: ‘__BIG_ENDIAN’
 static const bool kLittleEndian = !LEVELDB_IS_BIG_ENDIAN;
                                    ^~~~~~~~~~~~~~~~~~~~~
                                    __BIG_ENDIAN
Makefile:6638: recipe for target 'leveldb/db/leveldb_libleveldb_a-log_reader.o' failed
make[2]: *** [leveldb/db/leveldb_libleveldb_a-log_reader.o] Error 1
In file included from ./leveldb/port/port.h:14:0,
                 from ./leveldb/util/coding.h:18,
                 from ./leveldb/db/dbformat.h:17,
                 from leveldb/db/dbformat.cc:5:
./leveldb/port/port_stdcxx.h:44:36: error: ‘LEVELDB_IS_BIG_ENDIAN’ was not declared in this scope
 static const bool kLittleEndian = !LEVELDB_IS_BIG_ENDIAN;
                                    ^~~~~~~~~~~~~~~~~~~~~
...
```

**Comment by laanwj** [comment]
> I like the idea of a subtrees/thirdparty/whatever subdirectory, but not as part of this PR.

Same, I absolutely think it's a good idea, but not a good idea to commit to a specific way to doing it in this PR. It's classic scope creep. Now we'd have to agree on the directory scheme too.

> Edit: disregard, stupid mistake - I ran make clean but didn't run make distclean && ./autogen.sh && ./configure ....

Yep, be sure to build from a clean tree after large build system changes such as this one. Autotools dependency tracking is not good at picking up meta-changes.

**Comment by laanwj** [comment]
The new AppVeyor error is kind of interesting. All the RPC tests fail with:
```
 node1 2019-11-19T20:55:35.174988Z [] Opening LevelDB in C:\Users\appveyor\AppData\Local\Temp\1\test_runner_₿_🏃_20191119_205529\wallet_hd_132\node1\regtest\blocks\index 
 node1 2019-11-19T20:55:35.174988Z [] Fatal LevelDB error: NotFound: C:\Users\appveyor\AppData\Local\Temp\1\test_runner_₿_🏃_20191119_205529\wallet_hd_132\node1\regtest\blocks\index/LOCK: The system cannot find the path specified.\x0d 
 node1 2019-11-19T20:55:35.174988Z [] You can use -debug=leveldb to get more complete diagnostic messages 
 node1 2019-11-19T20:55:35.174988Z [] Fatal LevelDB error: NotFound: C:\Users\appveyor\AppData\Local\Temp\1\test_runner_₿_🏃_20191119_205529\wallet_hd_132\node1\regtest\blocks\index/LOCK: The system cannot find the path specified.\x0d 
 node1 2019-11-19T20:55:35.174988Z [] : Error opening block database. 
                                   Please restart with -reindex or -reindex-chainstate to recover.
```

**Comment by jamesob** [comment]
Benches finished. Performance looks basically equivalent, although it looks like leveldb 1.22 may use ~3% more memory, though this may just be noise.

leveldb-bench-plot

### bench/leveldb.1 vs. bench/leveldb.master.2 (absolute)
|                 bench name                 |  x  |       bench/leveldb.1       |  bench/leveldb.master.2   |
|--------------------------------------------|----:|-----------------------------|---------------------------|
| ibd.local.range.500000.535000.total_secs   |   2 | 4011.4344 (± 14.5325)       | 4022.3978 (± 12.0184)     |
| ibd.local.range.500000.535000.peak_rss_KiB |   2 | 1870192.0000 (± 37792.0000) | 1810804.0000 (± 596.0000) |
### bench/leveldb.1 vs. bench/leveldb.master.2 (relative)
|                 bench name                 |  x  | bench/leveldb.1 | bench/leveldb.master.2 |
|--------------------------------------------|----:|----------------:|-----------------------:|
| ibd.local.range.500000.535000.total_secs   |   2 |           1.000 |                  1.003 |
| ibd.local.range.500000.535000.peak_rss_KiB |   2 |           1.033 |                  1.000 |

**Comment by laanwj** [comment]
@jamesob Thanks for benchmarking! Which platform?

**Comment by jamesob** [comment]
Here's the host info: (I'll try to embed this in the graphs somehow)

| key                 | value                                     |
| -----------------   | ---------------------------------------   |
| hostname            | bench-strong                              |
| cpu_model_name      | Intel(R) Core(TM) i7-4770 CPU @ 3.40GHz   |
| ram_gb              | 31.353431701660156                        |
| os                  | ['Debian GNU/Linux', '10', 'buster']      |
| arch                | x86_64                                    |
| kernel              | 4.19.0-5-amd64                            |
| read_iops (/tmp)    | 27.5k                                     |
| write_iops (/tmp)   | 9215                                      |

**Comment by laanwj** [comment]
> The new AppVeyor error is kind of interesting. All the RPC tests fail with:
>
> Might have something to do with UTF-8 handling in paths. -D_UNICODE -DUNICODE was taken from the upstream build system into both the MSVC and autotools, but it's not clear it's enough.

Does anyone have an idea about windows unicode issues? (if this is even the reason for this) I remember there were some patches for that against the old windows environment (https://github.com/bitcoin-core/leveldb/commit/f8e797a058b7a3993314e985dfdff8124214ba99 I can find this quickly), but these can't be ported forward to the current one as-is.

File: Makefile.am [hunk 1]
```diff
@@ -189,7 +189,7 @@ $(BITCOIN_WALLET_BIN): FORCE
 	$(MAKE) -C src $(@F)
 
 if USE_LCOV
-LCOV_FILTER_PATTERN=-p "/usr/include/" -p "/usr/lib/" -p "src/leveldb/" -p "src/bench/" -p "src/univalue" -p "src/crypto/ctaes" -p "src/secp256k1"
+LCOV_FILTER_PATTERN=-p "/usr/include/" -p "/usr/lib/" -p "src/leveldb/" -p "src/crc32c" -p "src/bench/" -p "src/univalue" -p "src/crypto/ctaes" -p "src/secp256k1"
```
Comment (MarcoFalke) [inline Makefile.am]: (trivial conflict here)

**Comment by MarcoFalke** [comment]
Looks like this needs a new ubsan suppression `implicit-integer-sign-change:crc32c/*` in `test/sanitizer_supp...`

The newly added unit tests fails on windows: https://travis-ci.org/bitcoin/bitcoin/jobs/622714557#L3406 and https://ci.appveyor.com/project/DrahtBot/bitcoin/builds/29415262#L66

**Comment by laanwj** [comment]
Added the ubsan suppression, though I don't understand why this is suddenly needed where it wasn't before the rebase.

> The newly added unit tests fails on windows: https://travis-ci.org/bitcoin/bitcoin/jobs/622714557#L3406 and https://ci.appveyor.com/project/DrahtBot/bitcoin/builds/29415262#L66

Yes, this is issue https://github.com/google/leveldb/issues/755
Can't really move forward here until that is solved.

**Comment by MarcoFalke** [comment]
>  Can't really move forward here until that is solved.

It didn't sound like they are going to apply a patch. Unless maybe you can show a bug in chrome or so. They'd prefer we apply the patch to our own fork.

**Comment by laanwj** [comment]
Yes, they seem to have no interest in the issue.

Not sure how to move forward here, as there have been no demonstrated performance benefits here I may just close it. Rebasing this is kind of annoying as I have to re-import the trees every time.

**Comment by sipsorcery** [comment]
I was going to submit a PR to the leveldb repo. I suspect they'll reject it but if there's no rush maybe it'd be worth leaving this one open until that happens.

**Comment by laanwj** [comment]
> I was going to submit a PR to the leveldb repo. I suspect they'll reject it but if there's no rush maybe it'd be worth leaving this one open until that happens.

FWIW I'm happy to merge that in our local tree! I'd like to avoid the current situation of completely diverging from upstream again, but we still have three patches on top (of which it isn't clear whether they will ever make upsteam) anyway, what's one more?

**Comment by sipsorcery** [comment]
Here's the patch. Once it's applied the test from #17641 passes and my python functional tests got 3 out of 4 100% passes (25% failure rate on Windows is about normal).

[env_windows.cc.diff.txt](https://github.com/bitcoin/bitcoin/files/3952424/env_windows.cc.diff.txt)

[PR](https://github.com/google/leveldb/pull/760) also submitted to leveldb upstream repo.

**Comment by laanwj** [comment]
> Here's the patch. Once it's applied the test from #17641 passes and my python functional tests got 3 out of 4 100% passes (25% failure rate on Windows is about normal).

Thanks! Will apply it and rebase this.

**Comment by laanwj** [comment]
OK: Travis (including the new unit test on Windows) is passing now, AppVeyor failing same as it is everywhere.

**Comment by laanwj** [comment]
Rebased again (for trivial change in developer doc)

**Comment by sipsorcery** [comment]
Needs #17980 (and a few others) to be able to build again.

**Comment by MarcoFalke** [comment]
re-run ci

I am still planning to debug the gitian build failure

**Comment by laanwj** [comment]
@MarcoFalke a local gitian build also failed with
```
fatal: No url found for submodule path 'src/crc32c/third_party/benchmark' in .gitmodules
bin/gbuild:23:in `system!': failed to run cd inputs/bitcoin && git submodule update --init --recursive --force (RuntimeError)
        from bin/gbuild:304:in `block in <main>'
        from bin/gbuild:288:in `each'
        from bin/gbuild:288:in `<main>'
```
Looks like a conflict between `git subtree` and `git submodule`? I've imported `crc32c` as a subtree which (apparently) it has google benchmark as a submodule, which we entirely don't need, but gitian's checkout process notices it and fails.

If I understand this correctly, I think the most straightforward way this could be worked around is by removing the submodule in our [crc32c fork](https://github.com/bitcoin-core/crc32c) and subtreeing again.

**Comment by laanwj** [comment]
Anything left to do here? If not, can this get some (re-)ACKs please?

**Comment by sipsorcery** [comment]
ACK 66480821b36c839ab7615cb9309850015bceadb0.

No problems using a data directory of "₿_🏃" on Windows 10.

**Comment by sipa** [comment]
ACK 677fb8e92380d4deb6a3753047c01f7cf7b5af91

**Pull request merged**